execution/src/vm.rs

```diff
@@ -0,0 +1,298 @@
+            current_execution_context,
+        }
+    }
+
+    /// runs an SCE-final execution step
+    /// # Parameters
+    ///   * step: execution step to run
+    pub fn run_final_step(&mut self, step: ExecutionStep) {
+        // check if step already in history front
+        let opt_cached = {
+            if let Some((slot, opt_block, ledger_changes)) = self.step_history.pop_front() {
+                if slot != step.slot {
+                    // slot mismatch
+                    None
+                } else {
+                    match (&opt_block, &step.block) {
+                        (None, None) => {
+                            // matching miss
+                            Some(ledger_changes)
+                        }
+                        (Some(b_id_hist), Some((b_id_step, _b_step))) => {
+                            if b_id_hist == b_id_step {
+                                // matching block
+                                Some(ledger_changes)
+                            } else {
+                                // block mismatch
+                                None
+                            }
+                        }
+                        (None, Some(_)) => None, // miss/block mismatch
+                        (Some(_), None) => None, // block/miss mismatch
+                    }
+                }
+            } else {
+                None
+            }
+        };
+        if let Some(cached) = opt_cached {
+            // execution was already done, apply cached ledger changes to final ledger
+            let mut final_ledger_guard = self.final_ledger.lock().unwrap();
+            (*final_ledger_guard).apply_changes(&cached);
+            return;
+        }
+
+        // nothing found in cache, or cache mismatch: reset history, run step and make it final
+        // this should almost never happen, so the heavy step.clone() is OK
+        self.reset_to_final();
+        self.run_active_step(step.clone());
+        self.run_final_step(step);
+    }
+
+    /// runs an SCE-active execution step
+    /// # Parameters
+    ///   * step: execution step to run
+    pub fn run_active_step(&mut self, step: ExecutionStep) {
+        // accumulate active ledger changes history
+        let mut active_changes = SCELedgerChanges::default();
+        for (_step_slot, _step_opt_block, step_changes) in self.step_history.iter() {
+            active_changes.apply_changes(step_changes);
+        }
+
+        // setup active execution context
+        {
+            let mut exec_context_guard = self.current_execution_context.0.lock().unwrap();
+            // clear caused changes
+            (*exec_context_guard).ledger_step.caused_changes.0.clear();
+            // accumulated step history
+            (*exec_context_guard).ledger_step.cumulative_history_changes = active_changes;
+            // TODO add more info in the exec_context: slot, PoS draws, optional block, call stack etc...
```

Comment:
Why would you want the proof of stake here ?

Reply:
to allow the SCE access to info on the draws, for example to act on the coins/data of the block creator or endorsers, even if there is a miss. It also allows seeding a deterministic RNG that can be made available through a syscall.

Reply:
Can you link an issue to that todo

Reply:
done